plugins/micron: Added support for OCP telemetry (internal) log parsing. Datacenter NVMe SSD Specification v2.5r9 section 4.9

saskchaithanya opened the pull request:
**Description:**

- Added support for OCP telemetry (internal) log parsing as per OCP 2.5 specification - [Datacenter NVMe SSD Specification v2.5r9.pdf](https://github.com/user-attachments/files/16051920/Datacenter.NVMe.SSD.Specification.v2.5r9.pdf) section 4.9.

- The CLI command will take the input binary files (telemetry/internal logs LOG ID 0x07 or 0x08) and string log (0xC9) as inputs and provides a parsed JSON or text file as attached below:


**CLI Examples:**
$ sudo ./nvme micron ocp-telemetry-log-parse --format=normal --string-log="nvmelog_ocp_c9.bin" --telemetry-log="nvme_host_telemetry_log.bin" --output-file="nvme_cli_telemetry_host_normal.txt" /dev/nvme0

$ sudo ./nvme micron ocp-telemetry-log-parse --format=json --string-log="nvmelog_ocp_c9.bin" --telemetry-log="nvme_host_telemetry_log.bin" --output-file="nvme_cli_telemetry_host_normal.json" /dev/nvme0

$ sudo ./nvme micron ocp-telemetry-log-parse --format=normal --string-log="nvmelog_ocp_c9.bin" --telemetry-log="nvme_host_telemetry_log.bin" > nvme_cli_telemetry_host_console.txt /dev/nvme0

$ sudo ./nvme micron ocp-telemetry-log-parse --format=json --string-log="nvmelog_ocp_c9.bin" --telemetry-log="nvme_host_telemetry_log.bin" > nvme_cli_telemetry_host_console.json /dev/nvme0

**Output Files:**
[nvme_cli_telemetry_host_console.json](https://github.com/linux-nvme/nvme-cli/files/15432245/nvme_cli_telemetry_host_console.json)
[nvme_cli_telemetry_host_console.txt](https://github.com/linux-nvme/nvme-cli/files/15432246/nvme_cli_telemetry_host_console.txt)
[nvme_cli_telemetry_host_normal.json](https://github.com/linux-nvme/nvme-cli/files/15432250/nvme_cli_telemetry_host_normal.json)
[nvme_cli_telemetry_host_normal.txt](https://github.com/linux-nvme/nvme-cli/files/15432251/nvme_cli_telemetry_host_normal.txt)

igaw commented:
From a quick look, there are coding style issues, please see the output from checkpatch. It should help to identify the issues.

saskchaithanya commented:
> From a quick look, there are coding style issues, please see the output from checkpatch. It should help to identify the issues.

Fixed the coding style issues. Please review

igaw commented:
You should update the first commit instead of adding a new commit. The reason is that we follow the coding and submission rules from the linux kernel. Thus we are explicitly not allowing the github style submissions.

Also the commit message should be properly format and containing some information why you are changing thigns:

```
plugins/micron: Add support for OCP telemetry log parsing

Add support for OCP (internal) log parsing parsing. [maybe add the rerference to spec what part is implemented. This information should to help to review and also help to debug in future]

Signed-off-by: Chaithanya Shoba <ashoba@micron.com>
```

After looking at the changes, it contains several unrelated  changes in one PR. Please split them up, see the kernel documentation https://docs.kernel.org/process/submitting-patches.html#

The checkpatch.pl tool is very unhappy with this change:

```shell
$ ../linux/scripts/checkpatch.pl -g HEAD
[...]
ERROR: trailing whitespace
#5360: FILE: plugins/micron/micron-utils.h:48:
+ * @return integer value of hexadecimal $

ERROR: "foo* bar" should be "foo *bar"
#5372: FILE: plugins/micron/micron-utils.h:60:
+char* hex_to_ascii(const char *hex);

ERROR: "foo* bar" should be "foo *bar"
#5384: FILE: plugins/micron/micron-utils.h:72:
+unsigned char* read_binary_file(char *data_dir_path, const char *bin_path, long *buffer_size,

WARNING: line length of 104 exceeds 100 columns
#5400: FILE: plugins/micron/micron-utils.h:88:
+                                                       struct json_object *stats, __u8 spec, FILE *fp);

ERROR: Missing Signed-off-by: line by nominal patch author 'Chaithanya shoba <chaitu9es@gmail.com>'

total: 230 errors, 877 warnings, 5229 lines checked
```

Arunpandian15 commented:
@saskchaithanya 
Text documentation file is missing. Could you please add .txt file in documentation.

Arunpandian15 commented:
@saskchaithanya 
Please fix all the checkpatch review build failure. Most of them are related to coding style and compliance to the nvme-cli.

WARNING: Missing or malformed SPDX-License-Identifier tag in line 1
#6536: FILE: plugins/micron/micron-utils.h:1:
+// SPDX-License-Identifier: GPL-2.0-or-later
Error: WARNING: Missing or malformed SPDX-License-Identifier tag in line 1

ERROR: "foo* bar" should be "foo *bar"
#6565: FILE: plugins/micron/micron-utils.h:60:
+char* hex_to_ascii(const char *hex);
Error: ERROR: "foo* bar" should be "foo *bar"

ERROR: "foo* bar" should be "foo *bar"
#6574: FILE: plugins/micron/micron-utils.h:72:
+unsigned char* read_binary_file(char *data_dir_path, const char *bin_path, long *buffer_size,
Error: ERROR: "foo* bar" should be "foo *bar"

WARNING: line length of 104 exceeds 100 columns
#6585: FILE: plugins/micron/micron-utils.h:88:
+ struct json_object *stats, __u8 spec, FILE *fp);
Error: WARNING: line length of 104 exceeds 100 columns

ERROR: Missing Signed-off-by: line(s)
Error: ERROR: Missing Signed-off-by: line(s)

total: 200 errors, 773 warnings, 6430 lines checked

saskchaithanya commented:
Thanks @igaw and @Arunpandian15 for the review suggestions.
I am working on those and will update the pull request as early as possible.

igaw commented:
Please squash all changes into one patch, there is no need to see the development version of a change. And please no merges into the PR. Thanks.

ikegami-t commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+        return -1; // Invalid character
+}
+
+char* hex_to_ascii(const char* hex) 
```
```diff
- char* hex_to_ascii(const char* hex) 
+ char *hex_to_ascii(const char *hex) 
```

saskchaithanya replied:
corrected

ikegami-t commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+char* hex_to_ascii(const char* hex) 
+{
+    int hexLength = strlen(hex);
+    char* text = NULL;
```
```diff
-     char* text = NULL;
+     char *text = NULL;
```

saskchaithanya replied:
Fixed coding style issue.

ikegami-t commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+
+char* hex_to_ascii(const char* hex) 
+{
+    int hexLength = strlen(hex);
```
The camel case variable name hexLength should not be used as below. Also should be used the tab space indentation.
```diff
-     int hexLength = strlen(hex);
+ 	int hex_length = strlen(hex);
```

saskchaithanya replied:
Done

ikegami-t commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+    return text;
+}
+    
+unsigned char *read_binary_file(char *data_dir_path, const char *bin_path, long *buffer_size, int retry_count)
```
The line is tool long and also the function itself is too long so needed to split.

saskchaithanya replied:
splitted.

ikegami-t commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+    else
+    {
+        /* +2 for the / and null terminator */
+        file_path = (char *) calloc(1, strlen(data_dir_path) + strlen(bin_path) + 2);
```
```diff
-         file_path = (char *) calloc(1, strlen(data_dir_path) + strlen(bin_path) + 2);
+ 		file_path = (char *)calloc(1, strlen(data_dir_path) + strlen(bin_path) + 2);
```

saskchaithanya replied:
corrected

saskchaithanya commented:
[Datacenter NVMe SSD Specification v2.5r9.pdf](https://github.com/user-attachments/files/16051920/Datacenter.NVMe.SSD.Specification.v2.5r9.pdf)

Addressed the review comments and fixed the check-patch issues.
Please see the attached spec section 4.9 for reference and review.
Thanks!

saskchaithanya commented on plugins/micron/micron-nvme.c:
```diff
@@ -1,22 +1,12 @@
-#include "linux/types.h"
-#include "nvme-print.h"
-#include "util/cleanup.h"
+/* Copyright (c) Micron, Inc 2024.
```
Done as suggested.

saskchaithanya commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+
+
+void print_micron_vs_logs(__u8 *buf, struct micron_vs_logpage *log_page, int field_count,
+				 struct json_object *stats, __u8 spec, FILE *fp)
```
fp is used at the end of the function for writing to file. Please see at the end of the function.

saskchaithanya commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+		if (!size)
+			continue;
+		sfield = log_page[field].field;
+        if (size == 16) {
```
Thanks for the suggestion!
updated accordingly.

saskchaithanya commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+
+int hex_to_int(char c)
+{
+    if (c >= '0' && c <= '9')
```
Thanks for the link.
Updated the coding style accordingly.

saskchaithanya commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+    }
+}
+
+
```
Done

saskchaithanya commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+		else
+			printf("%-40s : %-4s\n", sfield, datastr);
+	}
+}
```
Added

saskchaithanya commented on plugins/micron/micron-utils.h:
```diff
@@ -0,0 +1,100 @@
+ * @return 0 success
+ */
+void print_micron_vs_logs(__u8 *buf, struct micron_vs_logpage *log_page, int field_count,
+				struct json_object *stats, __u8 spec, FILE *fp);
```
corrected the indentation.

saskchaithanya commented on plugins/micron/micron-utils.h:
```diff
@@ -0,0 +1,100 @@
+	char *field;
+	int  size;	/* FB client spec version 1.0 sizes - M5410 models */
+	int  size2; /* FB client spec version 0.7 sizes - M5407 models */
+}micron_vs_logpage_t, *pmicron_vs_logpage;
```
Thanks for the suggestion!
Followed the same convention as suggested.

saskchaithanya commented on plugins/micron/micron-utils.h:
```diff
@@ -0,0 +1,100 @@
+#include <limits.h>
+#include "linux/types.h"
+#include "nvme-print.h"
+#include "util/cleanup.h"
```
These headers are moved from micron-nvme.c. So needed.

saskchaithanya commented on plugins/micron/micron-utils.h:
```diff
@@ -0,0 +1,100 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/* Copyright (c) Micron, Inc 2024.
```
Done

saskchaithanya commented on plugins/micron/micron-utils.h:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @return pointer to binary data buffer
+ */
+unsigned char *read_binary_file(char *data_dir_path, const char *bin_path, long *buffer_size, int retry_count);
```
corrected

saskchaithanya commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+        else if(fp){
+            fprintf(fp, "%-40s : %-4s\n", sfield, datastr);
+        }			
+		else
```
Changed to switch case from if else for easy maintenance.

saskchaithanya commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,276 @@
+}
+
+
+void print_micron_vs_logs(__u8 *buf, struct micron_vs_logpage *log_page, int field_count,
```
May need to create a separate function for each case for splitting. so, changed to switch from if for easy maintenance. 
Please suggest if it is fine or still need to split. thanks

ikegami-t replied:
Still seems the switch case functions can be split more as below. Thank you.
```c
func_bbb()
{
...
}

func_***()
{
...
}

func_aaa()
{
    switch () {
    case bbb:
        func_bbb();
        break;
    ...
    default:
        func_***();
        break;
    }
}
```

saskchaithanya commented on plugins/micron/micron-utils.h:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @return 0 success
+ */
+void print_formatted_var_size_str(const char *msg, const __u8 *pdata, size_t data_size, FILE *fp);
```
Done

saskchaithanya commented:
> @saskchaithanya Please fix all the checkpatch review build failure. Most of them are related to coding style and compliance to the nvme-cli.
> 
> WARNING: Missing or malformed SPDX-License-Identifier tag in line 1 #6536: FILE: plugins/micron/micron-utils.h:1: +// SPDX-License-Identifier: GPL-2.0-or-later Error: WARNING: Missing or malformed SPDX-License-Identifier tag in line 1
> 
> ERROR: "foo* bar" should be "foo _bar" #6565: FILE: plugins/micron/micron-utils.h:60: +char_ hex_to_ascii(const char _hex); Error: ERROR: "foo_ bar" should be "foo *bar"
> 
> ERROR: "foo* bar" should be "foo _bar" #6574: FILE: plugins/micron/micron-utils.h:72: +unsigned char_ read_binary_file(char *data_dir_path, const char *bin_path, long _buffer_size, Error: ERROR: "foo_ bar" should be "foo *bar"
> 
> WARNING: line length of 104 exceeds 100 columns #6585: FILE: plugins/micron/micron-utils.h:88:
> 
> * struct json_object *stats, __u8 spec, FILE *fp);
>   Error: WARNING: line length of 104 exceeds 100 columns
> 
> ERROR: Missing Signed-off-by: line(s) Error: ERROR: Missing Signed-off-by: line(s)
> 
> total: 200 errors, 773 warnings, 6430 lines checked

Fixed the coding-style issues.

saskchaithanya commented:
> Please squash all changes into one patch, there is no need to see the development version of a change. And please no merges into the PR. Thanks.

Thanks a lot for the suggestion, Squashed the changes to one patch!

saskchaithanya commented:
> You should update the first commit instead of adding a new commit. The reason is that we follow the coding and submission rules from the linux kernel. Thus we are explicitly not allowing the github style submissions.
> 
> Also the commit message should be properly format and containing some information why you are changing thigns:
> 
> ```
> plugins/micron: Add support for OCP telemetry log parsing
> 
> Add support for OCP (internal) log parsing parsing. [maybe add the rerference to spec what part is implemented. This information should to help to review and also help to debug in future]
> 
> Signed-off-by: Chaithanya Shoba <ashoba@micron.com>
> ```
> 
> After looking at the changes, it contains several unrelated changes in one PR. Please split them up, see the kernel documentation https://docs.kernel.org/process/submitting-patches.html#
> 
> The checkpatch.pl tool is very unhappy with this change:
> 
> ```shell
> $ ../linux/scripts/checkpatch.pl -g HEAD
> [...]
> ERROR: trailing whitespace
> #5360: FILE: plugins/micron/micron-utils.h:48:
> + * @return integer value of hexadecimal $
> 
> ERROR: "foo* bar" should be "foo *bar"
> #5372: FILE: plugins/micron/micron-utils.h:60:
> +char* hex_to_ascii(const char *hex);
> 
> ERROR: "foo* bar" should be "foo *bar"
> #5384: FILE: plugins/micron/micron-utils.h:72:
> +unsigned char* read_binary_file(char *data_dir_path, const char *bin_path, long *buffer_size,
> 
> WARNING: line length of 104 exceeds 100 columns
> #5400: FILE: plugins/micron/micron-utils.h:88:
> +                                                       struct json_object *stats, __u8 spec, FILE *fp);
> 
> ERROR: Missing Signed-off-by: line by nominal patch author 'Chaithanya shoba <chaitu9es@gmail.com>'
> 
> total: 230 errors, 877 warnings, 5229 lines checked
> ```

Fixed check-patch issues and updated the commit message.

martin-belanger commented on plugins/micron/micron-nvme.c:
```diff
@@ -243,7 +234,7 @@ static int SetupDebugDataDirectories(char *strSN, char *strFilePath,
 			if (!fileLocation)
 				goto exit_status;
-			strncpy(fileLocation, strFilePath, length);
+			strscpy(fileLocation, strFilePath, sizeof(fileLocation));
```
Not sure you want to use `sizeof(fileLocation)` here since this is the size of the pointer and not the size of the memory pointed to by the pointer.

martin-belanger commented on plugins/micron/micron-nvme.c:
```diff
@@ -260,7 +251,7 @@ static int SetupDebugDataDirectories(char *strSN, char *strFilePath,
 					goto exit_status;
 				}
-				strcpy(strTemp, fileLocation);
+				strscpy(strTemp, fileLocation);
```
Doesn't `strscpy()` require a 3rd parameter (`count`)?

saskchaithanya replied:
This was an un-intended change, reverted back. Thanks for catching!

martin-belanger commented on plugins/micron/micron-nvme.c:
```diff
@@ -3169,7 +3079,7 @@ static int micron_internal_logs(int argc, char **argv, struct command *cmd,
 	}
 	sn[j] = '\0';
-	strcpy(ctrl.sn, sn);
+	strscpy(ctrl.sn, sn);
```
Missing `count`

saskchaithanya replied:
This is an un-intended change, reverted back. Thanks for catching!

saskchaithanya commented:
> @saskchaithanya Text documentation file is missing. Could you please add .txt file in documentation.

Working on this. I will add ASAP. Thanks!

igaw commented:
What are you editor settings? Are the tab settings set to 4 instead of 8? There are quiet a few places where the aliment is completely off:

![image](https://github.com/linux-nvme/nvme-cli/assets/1050803/e664c6b4-ca8d-4f33-a160-fbcfba1766f3)

ikegami-t commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,286 @@
+		case FIELD_SIZE_16: {
+			if (strstr(sfield, "GUID")) {
+				sprintf(datastr, "0x%"PRIx64"%"PRIx64"",
+						(uint64_t)le64_to_cpu(*(uint64_t *)(&buf[offset + 8])),
```
Is this casting really correct since seems below correct?
```diff
- (uint64_t)le64_to_cpu(*(uint64_t *)(&buf[offset + 8])),
+ le64_to_cpu(*(__u64 *)(&buf[offset + 8])),
```

saskchaithanya replied:
corrected, thanks.

ikegami-t commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,286 @@
+			} else if (strstr(sfield, "Timestamp")) {
+				lval_lo = *((__u64 *)(&buf[offset]));
+
+				char ts_buf[128];
```
char ts_buf definition seems better to move before lval_lo setting as same with the "Firmware Revision" else if case.

saskchaithanya replied:
Done

ikegami-t commented on plugins/micron/micron-utils.c:
```diff
@@ -0,0 +1,286 @@
+					sprintf(datastr, "0x%"PRIx64"", le64_to_cpu(lval_lo));
+			}
+		}
+		break;
```
The break and its indentation should be as below I think. (Basically I do not think good to use `{` and `}` for the each cases.)
```c
		switch (size) {
 		case FIELD_SIZE_16: {
			...
			break;
		}
		...
		}
```

saskchaithanya replied:
Done

ikegami-t commented on Documentation/meson.build:
```diff
@@ -80,6 +80,7 @@ adoc_sources = [
   'nvme-micron-selective-download',
   'nvme-micron-smart-add-log',
   'nvme-micron-temperature-stats',
+  'nvme-micron-ocp-telemetry-log-parse',
```
The ocp-telemetry-log-parse command documentation file is not added.

saskchaithanya replied:
Still working on this item.

saskchaithanya replied:
Added documentation file.

ikegami-t commented on completions/bash-nvme-completion.sh:
```diff
@@ -884,6 +884,9 @@ plugin_micron_opts () {
 		"vs-smbus-option")
 		opts+=" --option= -o --value= -v --save= -s"
 			;;
+		"ocp-telemetry-log-parse")
```
The zsh completions _nvme file is not added the command.

saskchaithanya replied:
Still working on this item.

saskchaithanya replied:
Added zsh completions _nvme file.

ikegami-t commented:
Still checkpatch.pl failed.

saskchaithanya commented:
> Still checkpatch.pl failed.

Thanks for checking. I will address the issues ASAP!

saskchaithanya commented:
> What are you editor settings? Are the tab settings set to 4 instead of 8? There are quiet a few places where the aliment is completely off:
> 
> ![image](https://private-user-images.githubusercontent.com/1050803/344716216-e664c6b4-ca8d-4f33-a160-fbcfba1766f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAyODU5MDAsIm5iZiI6MTcyMDI4NTYwMCwicGF0aCI6Ii8xMDUwODAzLzM0NDcxNjIxNi1lNjY0YzZiNC1jYThkLTRmMzMtYTE2MC1mYmNmYmExNzY2ZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDZUMTcwNjQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTNlYjdhYzdlYjY4MTg1NmNkZDBiZTAyYmNkNDc1MDFlNjBmMTI4ZmRmYjhjYTk1MDA5YTFjYmNjYjdkOWE1YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.fGwaM5V3Qc5xXqUWTw8lV3Y0LNSucpGs-Wu0kQIHI20)

Addressed the indentation issues. Thanks for checking!

saskchaithanya commented:
Could you please help running checkpatch.pl?

ikegami-t commented:
The script included by the linux kernel and can be cloned from the repository below.
```
git://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git 
```
Also the script can be executed as below for example.
```sh
tokunori@tokunori-desktop:~/nvme-cli$ ../linux/scripts/checkpatch.pl -g HEAD
```

saskchaithanya commented:
> The script included by the linux kernel and can be cloned from the repository below.
> 
> ```
> git://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git 
> ```
> 
> Also the script can be executed as below for example.
> 
> ```shell
> tokunori@tokunori-desktop:~/nvme-cli$ ../linux/scripts/checkpatch.pl -g HEAD
> ```

Thanks a lot, for the information!

HaroPanosyan commented on Documentation/nvme-micron-ocp-telemetry-log-parse.txt:
```diff
@@ -0,0 +1,89 @@
+-------
+-l <file>::
+--telemetry-log=<file>::
+	This option will allows the users to specify input telemetry-log file name.
```
This option will "allow", no s. 2 more "allows" below.

saskchaithanya replied:
Thanks for catching, corrected!

saskchaithanya commented:
Address the checkpatch issues and review comments. 
Please re-run the checkpatch and review latest changes.
Thanks!

igaw commented:
Almost there. There are some checkpatch complains left, nothing serious.

Though clang is not happy:

```
 FAILED: nvme.p/plugins_micron_micron-ocp-telemetry.c.o 
clang -Invme.p -I. -I.. -Iccan -I../ccan -Isubprojects/libnvme/src -I../subprojects/libnvme/src -I/usr/include/json-c -fcolor-diagnostics -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=gnu99 -O0 -g -fomit-frame-pointer -D_GNU_SOURCE -include config.h -DCCAN_LIST_DEBUG=1 -DCCAN_STR_DEBUG=1 -MD -MQ nvme.p/plugins_micron_micron-ocp-telemetry.c.o -MF nvme.p/plugins_micron_micron-ocp-telemetry.c.o.d -o nvme.p/plugins_micron_micron-ocp-telemetry.c.o -c ../plugins/micron/micron-ocp-telemetry.c
../plugins/micron/micron-ocp-telemetry.c:697:16: error: format string is not a string literal (potentially insecure) [-Werror,-Wformat-security]
                        fprintf(fp, buffer);
                                    ^~~~~~
../plugins/micron/micron-ocp-telemetry.c:697:16: note: treat the string as an argument to avoid this
                        fprintf(fp, buffer);
                                    ^
                                    "%s", 
../plugins/micron/micron-ocp-telemetry.c:699:11: error: format string is not a string literal (potentially insecure) [-Werror,-Wformat-security]
                        printf(buffer);
                               ^~~~~~
../plugins/micron/micron-ocp-telemetry.c:699:11: note: treat the string as an argument to avoid this
                        printf(buffer);
                               ^
                               "%s", 
```

You can trigger this build locally with 
```shell
$ scripts/build.sh -c clang
```

saskchaithanya commented:
> Almost there. There are some checkpatch complains left, nothing serious.
> 
> Though clang is not happy:
> 
> ```
>  FAILED: nvme.p/plugins_micron_micron-ocp-telemetry.c.o 
> clang -Invme.p -I. -I.. -Iccan -I../ccan -Isubprojects/libnvme/src -I../subprojects/libnvme/src -I/usr/include/json-c -fcolor-diagnostics -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=gnu99 -O0 -g -fomit-frame-pointer -D_GNU_SOURCE -include config.h -DCCAN_LIST_DEBUG=1 -DCCAN_STR_DEBUG=1 -MD -MQ nvme.p/plugins_micron_micron-ocp-telemetry.c.o -MF nvme.p/plugins_micron_micron-ocp-telemetry.c.o.d -o nvme.p/plugins_micron_micron-ocp-telemetry.c.o -c ../plugins/micron/micron-ocp-telemetry.c
> ../plugins/micron/micron-ocp-telemetry.c:697:16: error: format string is not a string literal (potentially insecure) [-Werror,-Wformat-security]
>                         fprintf(fp, buffer);
>                                     ^~~~~~
> ../plugins/micron/micron-ocp-telemetry.c:697:16: note: treat the string as an argument to avoid this
>                         fprintf(fp, buffer);
>                                     ^
>                                     "%s", 
> ../plugins/micron/micron-ocp-telemetry.c:699:11: error: format string is not a string literal (potentially insecure) [-Werror,-Wformat-security]
>                         printf(buffer);
>                                ^~~~~~
> ../plugins/micron/micron-ocp-telemetry.c:699:11: note: treat the string as an argument to avoid this
>                         printf(buffer);
>                                ^
>                                "%s", 
> ```
> 
> You can trigger this build locally with
> 
> ```shell
> $ scripts/build.sh -c clang
> ```

Thanks for the suggestion, Fixed the issues.

saskchaithanya commented on plugins/micron/micron-ocp-telemetry.c:
```diff
@@ -0,0 +1,1416 @@
+		if (fp)
+			fprintf(fp, buffer);
+		else
+			printf(buffer);
```
Fixed the issue.

saskchaithanya commented on plugins/micron/micron-ocp-telemetry.c:
```diff
@@ -0,0 +1,1416 @@
+
+		sprintf(buffer, "%s%s\n%s", STR_LINE, event_fifo_name, STR_LINE);
+		if (fp)
+			fprintf(fp, buffer);
```
Fixed the issue.

saskchaithanya commented:
Addressed the checkpatch issues and review comments.
Please re-run the checkpatch and review latest changes.
Thanks!

igaw commented:
Thanks!
